<div style="display:block">
    <div style="width: 20%; display: inline-block; text-align: left;">
    </div>
    <div style="width: 59%; display: inline-block">
        <h1  style="text-align: center">Assignment 2B</h1>
        <div style="width: 100%; text-align: center; display: inline-block;"><i>Author:</i> <strong>Pavan Kumar</strong> </div>
    </div>
    <div style="width: 20%; text-align: right; display: inline-block;">
        <div style="width: 100%; text-align: left; display: inline-block;">
            <i>Created: </i>
            <time datetime="2014-03-10" pubdate>18th May 2018</time>
        </div>
    </div>
</div>

In [1]:
import numpy as np
from IPython.display import display, Markdown, Latex

In [2]:
def sigmoid(x, derivative=False):
  return x * (1-x) if derivative else 1 /(1 + np.exp(-x))

## Step 0
Read the input and output
![Step 0](https://s3-ap-south-1.amazonaws.com/av-blog-media/wp-content/uploads/2017/05/26185640/0.0NN.jpg)

In [3]:
X = [[1, 0, 1, 0],
     [1, 0, 1, 1],
     [0, 1, 0, 1]]
Y = [[1, 1, 0]]

observed_data = np.array(X)
actual_dependent = np.array(Y)
print(observed_data)
print(actual_dependent)

[[1 0 1 0]
 [1 0 1 1]
 [0 1 0 1]]
[[1 1 0]]


## Step 1
Initialize weights and biases with random values (There are methods to initialize weights and biases but for now initialize with random values)

In [4]:
input_shape = np.shape(observed_data)
print(input_shape)
print("(Number of Rows, No. of Features)")

(3, 4)
(Number of Rows, No. of Features)


In [5]:
np.random.seed(666)

weights_h1 = np.around(np.random.rand(input_shape[1], input_shape[0]),2)

#We need 1 bias for each data point in the layer
bias_h1 = np.around(np.random.rand(1, input_shape[0]),2)
print("Weights - Hidden Layer 1")
print(weights_h1)
print("Bias - Hidden Layer 1")
print(bias_h1)


Weights - Hidden Layer 1
[[ 0.7   0.84  0.68]
 [ 0.73  0.95  0.01]
 [ 0.41  0.05  0.1 ]
 [ 0.51  0.2   0.74]]
Bias - Hidden Layer 1
[[ 0.19  0.7   0.29]]


In [6]:
output_shape = np.shape(actual_dependent)

In [7]:
weights_out = np.around(np.random.rand(output_shape[1], output_shape[0]),2)
bias_out = np.around(np.random.rand(1, output_shape[0]),2)
print("Weights - Output")
print(weights_out)
print("Bias - Output")
print(bias_out)

Weights - Output
[[ 0.77]
 [ 0.01]
 [ 0.11]]
Bias - Output
[[ 0.11]]


|X|  |  |  |wh  |     |     |bh  |     |     |hidden_layer_input|||hidden_layer_activations|||wout|bout|output|Y|E|
|----------|----------------|----------------|----|-----|---------|------|---------|---------|----|----|------|-|-|
|1| 0| 1| 0|0.70| 0.84| 0.68|0.19| 0.70| 0.29|    |     |         |      |         |         |0.37|0.22|      |1| |
|1| 0| 1| 1|0.73| 0.95| 0.01|    |     |     |    |     |         |      |         |         |0.08|    |      |1| |
|0| 1| 0| 1|0.41| 0.05| 0.10|    |     |     |    |     |         |      |         |         |0.78|    |      |0| |
| |  |  |  |0.51| 0.20| 0.74|    |     |     |    |     |         |      |         |         |    |    |      | | |

## Step 2
Calculate hidden layer input

In [8]:
hidden_layer_input = np.around(np.dot(observed_data, weights_h1) + bias_h1,2)

In [9]:
hidden_layer_input

array([[ 1.3 ,  1.59,  1.07],
       [ 1.81,  1.79,  1.81],
       [ 1.43,  1.85,  1.04]])

|X|  |  |  |wh  |     |     |bh  |     |     |hidden_layer_input|||hidden_layer_activations|||wout|bout|output|Y|E|
|----------|----------------|----------------|----|-----|---------|------|---------|---------|----|----|------|-|-|
|1| 0| 1| 0|0.70| 0.84| 0.68|0.19| 0.70| 0.29|1.3 |1.59 |1.07     |      |         |         |0.37|0.22|      |1| |
|1| 0| 1| 1|0.73| 0.95| 0.01|    |     |     |1.81|1.79 |1.81     |      |         |         |0.08|    |      |1| |
|0| 1| 0| 1|0.41| 0.05| 0.10|    |     |     |1.43|1.85 |1.04     |      |         |         |0.78|    |      |0| |
| |  |  |  |0.51| 0.20| 0.74|    |     |     |    |     |         |      |         |         |    |    |      | | |

# Step 3
Perform non-linear transformation on hidden linear input

In [10]:
hidden_layer_activations = np.around(sigmoid(hidden_layer_input),2)

In [11]:
hidden_layer_activations

array([[ 0.79,  0.83,  0.74],
       [ 0.86,  0.86,  0.86],
       [ 0.81,  0.86,  0.74]])

|X|  |  |  |wh  |     |     |bh  |     |     |hidden_layer_input|||hidden_layer_activations|||wout|bout|output|Y|E|
|----------|----------------|----------------|----|-----|---------|------|---------|---------|----|----|------|-|-|
|1| 0| 1| 0|0.70| 0.84| 0.68|0.19| 0.70| 0.29|1.3 |1.59 |1.07     |0.79  |0.83     |0.74     |0.37|0.22|      |1| |
|1| 0| 1| 1|0.73| 0.95| 0.01|    |     |     |1.81|1.79 |1.81     |0.86  |0.86     |0.86     |0.08|    |      |1| |
|0| 1| 0| 1|0.41| 0.05| 0.10|    |     |     |1.43|1.85 |1.04     |0.81  |0.86     |0.74     |0.78|    |      |0| |
| |  |  |  |0.51| 0.20| 0.74|    |     |     |    |     |         |      |         |         |    |    |      | | |

## Step 4
Perform linear and non-linear transformation of hidden layer activation at output layer

In [12]:
output_layer_input = np.dot(hidden_layer_activations, weights_out) + bias_out
output_layer_activation =  np.around(sigmoid(output_layer_input),2)

In [13]:
output_layer_activation

array([[ 0.69],
       [ 0.71],
       [ 0.7 ]])

|X|  |  |  |wh  |     |     |bh  |     |     |hidden_layer_input|||hidden_layer_activations|||wout|bout|output|Y|E|
|----------|----------------|----------------|----|-----|---------|------|---------|---------|----|----|------|-|-|
|1| 0| 1| 0|0.70| 0.84| 0.68|0.19| 0.70| 0.29|1.3 |1.59 |1.07     |0.79  |0.83    |0.74     |0.37|0.22|0.69  |1| |
|1| 0| 1| 1|0.73| 0.95| 0.01|    |     |     |1.81|1.79 |1.81     |0.86  |0.86     |0.86     |0.08|    |0.71  |1| |
|0| 1| 0| 1|0.41| 0.05| 0.10|    |     |     |1.43|1.85 |1.04     |0.81  |0.86     |0.74     |0.78|    |0.7   |0| |
| |  |  |  |0.51| 0.20| 0.74|    |     |     |    |     |         |      |         |         |    |    |      | | |

## Step 5
Calculate gradient of Error(E) at output layer

In [14]:
error = actual_dependent.T - output_layer_activation

In [15]:
error

array([[ 0.31],
       [ 0.29],
       [-0.7 ]])

|X|  |  |  |wh  |     |     |bh  |     |     |hidden_layer_input|||hidden_layer_activations|||wout|bout|output|Y|E|
|----------|----------------|----------------|----|-----|---------|------|---------|---------|----|----|------|-|-|
|1| 0| 1| 0|0.70| 0.84| 0.68|0.19| 0.70| 0.29|1.3 |1.59 |1.07     |0.79  |0.83     |0.74    |0.37|0.22|0.69  |1|0.31|
|1| 0| 1| 1|0.73| 0.95| 0.01|    |     |     |1.81|1.79 |1.81     |0.86  |0.86     |0.86     |0.08|    |0.71  |1|0.29|
|0| 1| 0| 1|0.41| 0.05| 0.10|    |     |     |1.43|1.85 |1.04     |0.81  |0.86     |0.74     |0.78|    |0.7   |0|-0.7|
| |  |  |  |0.51| 0.20| 0.74|    |     |     |    |     |         |      |         |         |    |    |      | | |

## Step 6
Compute slope at output and hidden layer

In [16]:
slope_output_activation = np.around(sigmoid(output_layer_activation, derivative=True),2)

In [17]:
slope_output_activation

array([[ 0.21],
       [ 0.21],
       [ 0.21]])

In [18]:
slope_hidden_layer = np.around(sigmoid(hidden_layer_activations, derivative=True),2)

In [19]:
slope_hidden_layer

array([[ 0.17,  0.14,  0.19],
       [ 0.12,  0.12,  0.12],
       [ 0.15,  0.12,  0.19]])

|X|  |  |  |wh  |     |     |bh  |     |     |hidden_layer_input|||hidden_layer_activations|||wout|bout|output|Y|E|
|----------|----------------|----------------|----|-----|---------|------|---------|---------|----|----|------|-|-|
|1| 0| 1| 0|0.70| 0.84| 0.68|0.19| 0.70| 0.29|1.3 |1.59 |1.07     |0.79  |0.83     |0.74    |0.37|0.22|0.69  |1|0.31|
|1| 0| 1| 1|0.73| 0.95| 0.01|    |     |     |1.81|1.79 |1.81     |0.86  |0.86     |0.86     |0.08|    |0.71  |1|0.29|
|0| 1| 0| 1|0.41| 0.05| 0.10|    |     |     |1.43|1.85 |1.04     |0.81  |0.86     |0.74     |0.78|    |0.7   |0|-0.7|
| |  |  |  |0.51| 0.20| 0.74|    |     |     |    |     |         |      |         |         |    |    |      | | |


|slope_hidden_layer|    |     |slope_output|
|------------------|----|-----|------------|
|0.17              |0.14|0.19 |0.21        |
|0.12              |0.12|0.12 |0.21        |
|0.15              |0.12|0.19 |0.21        |

# Step 7
Compute delta at output layer

In [20]:
delta_output = np.around(slope_output_activation * error,2)

In [21]:
delta_output

array([[ 0.07],
       [ 0.06],
       [-0.15]])

|X|  |  |  |wh  |     |     |bh  |     |     |hidden_layer_input|||hidden_layer_activations|||wout|bout|output|Y|E|
|----------|----------------|----------------|----|-----|---------|------|---------|---------|----|----|------|-|-|
|1| 0| 1| 0|0.70| 0.84| 0.68|0.19| 0.70| 0.29|1.3 |1.59 |1.07     |0.79  |0.83     |0.74    |0.37|0.22|0.69  |1|0.31|
|1| 0| 1| 1|0.73| 0.95| 0.01|    |     |     |1.81|1.79 |1.81     |0.86  |0.86     |0.86     |0.08|    |0.71  |1|0.29|
|0| 1| 0| 1|0.41| 0.05| 0.10|    |     |     |1.43|1.85 |1.04     |0.81  |0.86     |0.74     |0.78|    |0.7   |0|-0.7|
| |  |  |  |0.51| 0.20| 0.74|    |     |     |    |     |         |      |         |         |    |    |      | | |


|slope_hidden_layer|    |     |slope_output| Error |
|------------------|----|-----|------------|-------|
|0.17              |0.14|0.19 |0.21        |0.31|
|0.12              |0.12|0.12 |0.21        |0.29|
|0.15              |0.12|0.19 |0.21        |-0.7|

|delta_output|
|------------|
| 0.07|
| 0.06|
|-0.15|

## Step 8
Calculate Error at hidden layer

In [22]:
error_h1 = np.around(np.dot(delta_output, weights_out.T),2)

In [23]:
error_h1

array([[ 0.05,  0.  ,  0.01],
       [ 0.05,  0.  ,  0.01],
       [-0.12, -0.  , -0.02]])

|X|  |  |  |wh  |     |     |bh  |     |     |hidden_layer_input|||hidden_layer_activations|||wout|bout|output|Y|E|
|----------|----------------|----------------|----|-----|---------|------|---------|---------|----|----|------|-|-|
|1| 0| 1| 0|0.70| 0.84| 0.68|0.19| 0.70| 0.29|1.3 |1.59 |1.07     |0.79  |0.83     |0.74    |0.37|0.22|0.69  |1|0.31|
|1| 0| 1| 1|0.73| 0.95| 0.01|    |     |     |1.81|1.79 |1.81     |0.86  |0.86     |0.86     |0.08|    |0.71  |1|0.29|
|0| 1| 0| 1|0.41| 0.05| 0.10|    |     |     |1.43|1.85 |1.04     |0.81  |0.86     |0.74     |0.78|    |0.7   |0|-0.7|
| |  |  |  |0.51| 0.20| 0.74|    |     |     |    |     |         |      |         |         |    |    |      | | |



|slope_hidden_layer|    |     |error at hidden layer|  |     |slope_output| Error |
|------------------|----|-----|---------------------|--|-----|------------|-------|
|0.17              |0.14|0.19 |0.05                 | 0|0.01 |0.21        |0.31|
|0.12              |0.12|0.12 |0.05                 | 0|0.01 |0.21        |0.29|
|0.15              |0.12|0.19 |-0.12                | 0|-0.02|0.21        |-0.7|



|delta_output|
|------------|
| 0.07|
| 0.06|
|-0.15|

# Step 9 
Compute delta at hidden layer

In [24]:
delta_h1 = np.dot(error_h1, slope_hidden_layer)

In [25]:
delta_h1

array([[ 0.01  ,  0.0082,  0.0114],
       [ 0.01  ,  0.0082,  0.0114],
       [-0.0234, -0.0192, -0.0266]])


|slope_hidden_layer|    |     |error at hidden layer|  |     |slope_output| Error |
|------------------|----|-----|---------------------|--|-----|------------|-------|
|0.17              |0.14|0.19 |0.05                 | 0|0.01 |0.21        |0.31|
|0.12              |0.12|0.12 |0.05                 | 0|0.01 |0.21        |0.29|
|0.15              |0.12|0.19 |-0.12                | 0|-0.02|0.21        |-0.7|

|delta hidden layer|  |  |
|------------------|--|--|
|0.01              |0.0082 |0.0114 |
|0.01              |0.0082 |0.0114 |
|-0.0234           |-0.0192|-0.0266|

## Step 10
Update weight at both output and hidden layer


In [26]:
learning_rate = 0.1

In [27]:
weights_out =  weights_out + np.dot(hidden_layer_activations.T , delta_output) * learning_rate

In [28]:
weights_out

array([[ 0.76854],
       [ 0.00807],
       [ 0.10924]])

In [29]:
weights_h1 = weights_h1 + np.dot(observed_data.T, delta_h1) * learning_rate

In [30]:
weights_h1

array([[ 0.702  ,  0.84164,  0.68228],
       [ 0.72766,  0.94808,  0.00734],
       [ 0.412  ,  0.05164,  0.10228],
       [ 0.50866,  0.1989 ,  0.73848]])

## Step 11
Update biases at both output and hidden layer



In [32]:
bias_out = bias_out + np.sum(delta_output, axis=0) * learning_rate

In [33]:
bias_out

array([[ 0.108]])

In [34]:
bias_h1 = bias_h1 + np.sum(delta_h1, axis=0) * learning_rate

In [35]:
bias_h1

array([[ 0.18966,  0.69972,  0.28962]])

|X|  |  |  |wh  |     |     |bh  |     |     |hidden_layer_input|||hidden_layer_activations|||wout|bout|output|Y|E|
|----------|----------------|----------------|----|-----|---------|------|---------|---------|----|----|------|-|-|
|1| 0| 1| 0|0.702| 0.842| 0.682|0.189| 0.699| 0.289|1.3 |1.59 |1.07     |0.79  |0.83     |0.7    |0.768|0.22|0.69  |1|0.31|
|1| 0| 1| 1|0.728| 0.948| 0.007|    |     |     |1.81|1.79 |1.81     |0.86  |0.86     |0.86     |0.008|    |0.71  |1|0.29|
|0| 1| 0| 1|0.412| 0.052| 0.102|    |     |     |1.43|1.85 |1.04     |0.81  |0.86     |0.74     |0.109|    |0.7   |0|-0.7|
| |  |  |  |0.508| 0.199| 0.738|    |     |     |    |     |         |      |         |         |    |    |      | | |
